### DISCUSSION TITLE
Difference between webgl1 and webgl2

### DISCUSSION THREAD

**Discussion by goldenratio** [pixijs/pixijs#8631] [Q&A] [answered]
Hey,

is there any write up on differences between webgl1 and webgl2 from Pixi.js perspective.

**Comment by bigtimebuddy** [comment, marked as answer]
There should be an in-depth article about this, but I think the reason why there isn't is there isn't a huge difference with respect to Pixi between WebGL1 and WebGL2. When we added support for WebGL2, it did not really offer a huge set of benefits performance or feature-wise, so most of the WebGL2 bits are somewhat under-the-hood optimizations. 

That said, there are a few features to call out:

### Better Support for Non-Power of Two (NPOT) Textures

WebGL2, in general, has better support for NPOT textures. There are some places where this can be beneficial, like using the `WRAP_MODE.CLAMP` with NPOT textures. Specifically, there are some objects like TilingSprite, SimpleRope, NineSlicePlane that are more flexible in WebGL2 with NPOT textures.

### Multisampling (MSAA)

Multisampling is an experimental feature that requires WebGL2. It allows the use of antialiasing in individual renderTextures. Antialiasing is the same as the main buffer with renderer `antialias:true` option. MSAA has serious GPU memory consumption and GPU performance so it's not really something you'd enable across a large project or want to do every frame. But if you're trying to render high-quality Graphics, for instance, with antialias turn off this is a great feature.

### Unsigned 32-Bit Indices

WebGL2 can support 32-bit indices. Basically, this means much larger more complex Graphics and Mesh objects. In practice, this doesn't come up a lot unless you're really rendering content at the margins of complexity.

### Additional Texture Formats

WebGL2 supports a [broader variety of texture formats](https://github.com/pixijs/pixijs/blob/dev/packages/core/src/textures/utils/mapTypeAndFormatToInternalFormat.ts). For example: `gl.RGBA8UI`, `gl.DEPTH32F_STENCIL8`, `gl.RGB9_E5`, `gl.R11F_G11F_B10F`, etc. For users working more closely with different types of image buffers and formats with advanced use-cases, this can be helpful. For the vast majority of users, these aren't that applicable.
